Add HLL numerical flux (#648)
illinois-ceesd/mirgecom: doc/discretization.rst

Numerical fluxes
================

Numerical fluxes are responsible for calculating the fluxes at the DG element boundaries.  
Numerical fluxes must account for the discontinuities at element faces, and calculate
a single valued flux that both elements agree on.  That is, they must be functions

> of both $\pm$ states, and must produce a consistent flux. In general, numerical
> fluxes for the divergence operator must satisfy the consistency relations:

Reviewer: ```suggestion
of both $\pm$ states, and must produce a consistent flux.

For a conservation law $\frac{\partial \b{Q}}{\partial t} + \nabla \cdot \b{F}(\b{Q}) = \b{S}$,
the numerical flux $h$ must satisfy the consistency relations
```

.. math::
   h(\b{Q}, \b{Q}; \b{n}) = \b{F}(\b{Q})\cdot\b{n}~~~~~~

>    h(\b{Q}^+,\b{Q}^-;\b{n}) = -h(\b{Q}^-, \b{Q}^+;-\b{n}),

Reviewer: ```suggestion
   h(\b{Q}^+,\b{Q}^-;\b{n}) = -h(\b{Q}^-, \b{Q}^+;-\b{n}).
```

> where $h$ is the numerical flux function, and $\b{F}$ is the flux function at
> the element boundary the numerical flux function is intended to replace. 

Reviewer: ```suggestion
```

:$\b{Q}^{\pm}:$: Conserved quantities on the exterior/interior of the boundary face
:$\b{Q}_{bc}:$: Boundary condition for the fluid conserved quantities
:$\b{\Sigma}^{\pm}:$: Gradient of conserved quantities on ext/int of boundary face
:$\b{\Sigma}_{bc}:$: Boundary condition for gradient of the conserved quantities
:$\b{v}^{\pm}:$: Flow velocity on the exterior/interior of boundary face
:$h^*_e:$: Boundary facial flux for the divergence of the inviscid flux
:$h^*_v:$: Boundary facial flux for divergence of viscous flux
:$\b{H}_s^*:$: Boundary flux vector for the gradient of the conserved quantities
:$\hat{\b{n}}:$: Outward pointing unit normal for the boundary face

For all $\partial E \cap \partial\Omega$ the $+$ side is on the domain boundary. 
Boundary conditions ($\b{Q}_{bc}, \b{\Sigma}_{bc}$) are set by prescribing one or more
components of the solution or its gradient on the (+) side of the boundary,
($\b{Q}^+, \b{\Sigma}^+$), respectively, or by prescribing one or more components of the
boundary fluxes $h^*_e$, $h^*_v$, and $\b{H}^*_s$.  Descriptions of particular
boundary treatments follow in the next few sections.

> .. image:: figures/ElementBoundary.png

Author: fwiw, we intend to replace this outright with tikz.

> .. image:: figures/ElementBoundary.png

Author: #667

Numerical fluxes
================

Numerical fluxes are responsible for calculating the fluxes at the DG element boundaries.  
Numerical fluxes must account for the discontinuities at element faces, and calculate
a single valued flux that both elements agree on.  That is, they must be functions
of both $\pm$ states, and must produce a consistent flux. In general, numerical
fluxes for the divergence operator must satisfy the consistency relations:

.. math::
   \b{h}(\b{Q}, \b{Q}; \b{n}) = \b{F}(\b{Q})\cdot\b{n}~~~~~~
   \b{h}(\b{Q}^+,\b{Q}^-;\b{n}) = -\b{h}(\b{Q}^-, \b{Q}^+;-\b{n}),

where $\b{h}$ is the numerical flux function, and $\b{F}$ is the flux function at

> the element boundary the numerical flux function is intended to replace. 

Author: I'd be happy to if I could find them stated somewhere.  The only place I've seen these consistency relations stated, they are in ref. to the discussion of the Riemann solver based fluxes.

> the element boundary the numerical flux function is intended to replace. 

Reviewer: Hmm, unfortunately I don't know them either. Do you have a link to where you found the ones for div? Maybe we can derive them analogously.

My guess is they should be the same for the gradient (with dot replaced by multiplication). AFAICT the first ensures that the numerical flux agrees with the physical flux when the two sides agree on the state, and the second ensures that the surface integrals computed by adjacent faces agree (up to a minus sign due to difference in orientation).

> the element boundary the numerical flux function is intended to replace. 

Author: Eqn 10 of the [BR1 reference](https://www.sciencedirect.com/science/article/pii/S0021999196955722?via%3Dihub)
There's some discussion of the underpinning of this in 6.6.2 of my HW, Revisiting Numerical Fluxes or some such.

> the element boundary the numerical flux function is intended to replace. 

Reviewer: Meh, I'm not sure if it's worth the trouble. I think I'll just mark this as resolved and possibly suggest a minor change in wording below.

\b{H}_s(\b{Q}^+_h, \b{Q}_h^-; \b{n}) &\approx \b{Q}_h\b{n}.
$$

The $\b{h}_e$, and $\b{h}_v$ numerical flux functions are responsible for
calculating the numerical fluxes for the divergence of the inviscid and viscous
transport fluxes, respectively.  The $\b{H}_s$ numerical flux function is
responsible for calculating the numerical flux for the gradient operator, and

> so is employed, for example, in the auxiliary equation.

Author: Are you sure?  Can I reword it to make it better for you?  This is the only place where the actual notation for the numerical fluxes used in the rest of the document are actually connected with the physical flux terms they are intended to replace.

> so is employed, for example, in the auxiliary equation.

Reviewer: This paragraph in the previous section seems sufficient to me:

![Screen Shot 2022-05-10 at 12 15 53 PM](https://user-images.githubusercontent.com/7232918/167685528-3a36c131-63d1-48e4-9ca1-334a1b030890.png)

> so is employed, for example, in the auxiliary equation.

Author: Those are the internal face numerical flux functions.  The fluxes under discussion are those used at the domain boundary.  They are not the same functions, in general, and are documented as separate flux treatments in both BR1 and Mengaldo references.

> so is employed, for example, in the auxiliary equation.

Author: Removed.

Gradient numerical flux
^^^^^^^^^^^^^^^^^^^^^^^
The numerical flux function used for the gradient of the fluid solution,
$\b{H}_s$, is implemented in :func:`~mirgecom.flux.gradient_flux`
as follows:

.. math::
   \b{H}_s(\b{Q}_h^+, \b{Q}_h^- ; \b{n}) = \frac{1}{2}\left(
   (\b{Q}_h^+ + \b{Q}_h^-) \mp \beta(\b{Q}_h^+ - \b{Q}_h^-))\right)\b{n},

where again, $\beta$ is an optional parameter to add a dissipative term
to the gradient fluxes. $\beta$ defaults to $0$, resulting in a centered
numerical flux for the gradient.

.. note::
   The dissipative term coefficient $\beta$ used for the viscous numerical
   flux fucntion, $\b{h}_v$, and for the gradient numerical flux function,
   $\b{H}_s$, should be equal and opposite.

Domain boundary treatments
==========================

What happens when $\partial E \cap \partial\Omega \neq \emptyset$?

In DG, fluxes are responsible for handling the flow of information
between adjacent cells, and for transferring the boundary conditions
into the domain from the domain boundary.  In this sense, all of the
boundary conditions are *weak*, in that they are weakly enforced
through the flux, as opposed to directly setting the solution of the

> boundary elements (a *strong* enforcement).

Author: This point is covered by the paper that we ended up using to implement the BCs.  I think we are talking about the same thing - SAT is just another approach of weak boundary conditions.  I think they are weak by virtue of not prescribing the BC directly, but by providing the _flux_ that drives the soln to the BC.

> boundary elements (a *strong* enforcement).

Author: I re-worded this.  Can you take another look?

> boundary elements (a *strong* enforcement).

Reviewer: 🤷‍♂️ I guess I can live with it. I'm not really convinced yet that what the Mengaldo paper describes as "Weak-Prescribed" is any different numerically from strong enforcement... but I would have to read the paper more carefully to say for sure.

Boundary treatments in *MIRGE-Com* follow the prescriptions of the
the so-called BR1 method descibed by [Bassi_1997]_, and the boundary
treatment strategies outlined by [Mengaldo_2014]_.

The relevant quantities for the boundary treatments are as follows:

.. math::
   
  \b{Q}^{\pm} &\equiv \text{conserved quantities on the exterior/interior of the boundary face} \\
  \b{Q}_{bc} &\equiv \text{boundary condition for the fluid conserved quantities} \\
  \b{\Sigma}^{\pm} &\equiv \text{gradient of conserved quantities on ext/int of boundary face} \\
  \b{\Sigma}_{bc} &\equiv \text{boundary condition for grad of soln } \\
  \b{v}^{\pm} &\equiv \text{Flow velocity on the ext/int of boundary face} \\
  \b{h}^*_e &\equiv \text{boundary flux for the divergence of inviscid flux} \\
  \b{h}^*_v &\equiv \text{boundary flux for divergence of viscous flux} \\
  \b{H}_s^* &\equiv \text{boundary flux for the gradient of the conserved quantities} \\

>   \hat{\b{n}} &\equiv \text{outward pointing normal for the boundary face}

Reviewer: Maybe something like:

```
$\b{Q}^{\pm}$:
    conserved quantities on the exterior/interior of the boundary face
$\b{Q}_{bc}$:
    boundary condition for the fluid conserved quantities
$\b{\Sigma}^{\pm}$:
    gradient of conserved quantities on ext/int of boundary face
$\b{\Sigma}_{bc}$:
    boundary condition for grad of soln
$\b{v}^{\pm}$:
    flow velocity on the ext/int of boundary face
$\b{h}^*_e$:
    boundary flux for the divergence of inviscid flux
$\b{h}^*_v$:
    boundary flux for divergence of viscous flux
$\b{H}_s^*$:
    boundary flux for the gradient of the conserved quantities
$\hat{\b{n}}$:
    outward pointing normal for the boundary face
```

![Screen Shot 2022-05-10 at 10 52 39 AM](https://user-images.githubusercontent.com/7232918/167670695-629e6656-dcf2-467f-a19a-f43964d761f3.png)
?

>   \hat{\b{n}} &\equiv \text{outward pointing normal for the boundary face}

Author: I found `definition list` in RST format which rendered almost identical to above.  Whadda ya think?

For all $\partial E \cap \partial\Omega$ the $+$ side is on the domain boundary. 
Boundary conditions ($\b{Q}_{bc}, \b{\Sigma}_{bc}$) are set by prescribing one or more
components of the solution or its gradient on the (+) side of the boundary,
($\b{Q}^+, \b{\Sigma}^+$), respectively, or by prescribing one or more components of the
boundary fluxes $\b{h}^*_e$, $\b{h}^*_v$, and $\b{H}^*_s$.  Descriptions of particular
boundary treatments follow in the next few sections.

> .. image:: figures/ElementBoundary.png

Reviewer 2: Track down the source file?

Reviewer 2: https://sphinxcontrib-tikz.readthedocs.io/en/latest/

Author: > Track down the source file?

I have it, if it is useful.  PPTX

Reviewer 2: #667

   :width: 300
   :alt: Quantities at element boundary.
   :align: center

> Adiabtic slip wall

Reviewer 2: Read to here

with fluid velocity $\b{v}_{fluid}$, wall velocity $\b{v}_{wall}$, and outward pointing
unit normal $\hat{\b{n}}$. For a fixed wall, $\b{v}_{fluid} \cdot \hat{\b{n}} = 0$. The
components of the fluid velocity in the plane of the wall are left unperturbed by the wall.

More specifically, for the fixed wall in |mirgecom|, the fluid solution corresponding to this
boundary condition is this:

.. math::
   
   \b{Q}_{bc} = \begin{bmatrix}\rho^{-}\\(\rho{E})^{-}\\(\rho{v_b})_{i}\\(\rho{Y})^{-}_{\alpha}\end{bmatrix},

where $\mathbf{v}_b = \mathbf{v}^{-} - (\mathbf{v}^{-}\cdot\hat{\mathbf{n}})\hat{\mathbf{n}}$.

In |mirgecom|, this boundary condition is transmitted to the boundary element through the
approximate Riemann solver, or numerical flux function, $h_e(\b{Q}^-, \b{Q}^+)$.  As such,
the boundary treatment in |mirgecom| is to prescribe the boundary solution $\b{Q}^+$ to be
used in the numerical flux function to *induce* the desired boundary condition, $\b{Q}_{bc}$.

The adiabatic slip wall boundary treatment is implemented by the
:class:`~mirgecom.boundary.AdiabaticSlipBoundary`. The boundary solution
is prescribed as follows:

.. math::

>    \b{Q}^+ = \b{Q}^- - 2*\left(\rho\b{v}^-\cdot\hat{\b{n}}\right)\hat{\b{n}},

Author: Fixed.

   that is distinct from that used for the advective terms. This reference recommends
   explicitly setting the boundary velocities to zero for the $\b{Q}_{bc}$ used in
   computing $\nabla{\b{Q}}$ and $\b{F}_v(\b{Q}_{bc})$.

BR1 and Mengaldo prescribe the following boundary treatment:

The viscous boundary flux at solid walls is computed as:

.. math::

   \b{h}^*_v(\b{Q}_{bc}, \b{\Sigma}_{bc}) = \b{F}_V(\b{Q}_{bc},\b{\Sigma}_{bc})\cdot\b{n},

where $\b{Q}_{bc}$ are the same values used to prescribe $\b{h}^*_e$.

If there are no conditions on $\nabla\b{Q}\cdot\b{n}$, then:
$$
\b{\Sigma}_{bc} = \b{\Sigma}_h^-.
$$
Otherwise, $\b{\Sigma}_{bc}$ will need to be modified accordingly.

--------

MIRGE-Com currently does the following:

.. math::

   \b{h}^*_v(\b{Q}_{bc}, \b{\Sigma}_{bc}) = \b{h}_v\left(\b{Q}^-,\b{\Sigma}^-,\b{Q}_{bc},\b{\Sigma}_{bc}\right),

where $\b{Q}_{bc}$ are the same values used to prescribe $\b{h}^*_e$.

In *MIRGE-Com*, we use the central flux to transfer viscous BCs to the domain:

.. math::

     \b{h}^*_v(\b{Q}_{bc}, \b{\Sigma}_{bc}) = \frac{1}{2}\left(\mathbf{F}_v(\mathbf{Q}_{bc},\mathbf{\Sigma}_{bc}) + \mathbf{F}_v(\mathbf{Q}^{-},\mathbf{\Sigma}^{-})\right)

> 

Author: You are correct.  We implemented _new_ boundary conditions to match Mengaldo.  This document describes what is done in the BCs that are in the code at this point (i.e., in _this_ PR).  I'll update the docs to the new BCs in a separate PR.
